### PR TITLE
Benchmarking

### PR DESCRIPTION
**Pull request by jlavelle**

### CODE REVIEW

File: bench/index.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+const util = require("./util");
+const mini = require("minimist");
+const { adt } = require("@masaeedu/adt");
+const { Obj, Maybe, Either, Arr } = require("../src");
+const { Just, Nothing } = Maybe;
+const { Left, Right } = Either;
+
+const Command = adt({
+  RunAll: ["p"],
+  RunSuite: ["n", "s", "p"],
```
Comment (masaeedu) [inline bench/index.js]: Is this supposed to hold some specific types or just arbitrary things?

File: bench/index.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  List: ["_"],
+  Help: ["_"]
+});
+
+const { RunAll, RunSuite, RunSingle, List, Help } = Command;
+
+const parseCommand = args => {
+  const parsedArgs = (() => {
+    const named = Obj.map(k => {
+      const r = Obj.get(k)(args);
```
Comment (masaeedu) [inline bench/index.js]: Maybe `Obj.get` should just return a `Maybe`, but that's for a different PR.

File: bench/index.js (hunk 1 above)
Comment (jlavelle) [reply bench/index.js]: Specific things; addressed in 6c5a0e4.

File: bench/index.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    case "help":
+      return Right(Help());
+    case undefined:
+      return Right(RunAll(Nothing));
+    default:
+      return Left("Unknown mode " + args._[0]);
+  }
+};
+
+const maybeSaveResults = msavePath => suites =>
```
Comment (masaeedu) [inline bench/index.js]: You can also do:

```
const maybeSaveResults = Maybe.match({
  Just: suites => p => util.saveResults(p, suites),
  Nothing: _ => Nothing
})
```

btw. Not sure if that's better in this situation.

File: bench/index.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    Just: Right,
+    Nothing: Left("Missing name argument")
+  })(parsedArgs.name);
+
+  const parseSingleName = Either.chain(x => {
+    const [a, b] = x.split(".");
+    return a && b ? Right([a, b]) : Left(`invalid benchmark path ${x}`);
+  })(parseName);
+
+  const parseSizes = Right(
```
Comment (masaeedu) [inline bench/index.js]: Is this a `Maybe (Maybe something)`?

File: bench/index.js (hunk 3 above)
Comment (jlavelle) [reply bench/index.js]: I did end up rearranging the arguments  for this.

File: bench/index.js (hunk 4 above)
Comment (jlavelle) [reply bench/index.js]: It's an `Either String (Maybe [Int])`, I changed it a bit in d339848 so that it actually makes sure it has a list of `Int`s.

**Pull request merged**